doc/keynames/README.md

```diff
@@ -0,0 +1,477 @@
+| `/elektra/./../version` | `/elektra/version` |
+| `/elektra/version/`     | `/elektra/version` |
+
+So far this mostly follows, what we know from UNIX paths.
+However, the last two examples are somewhat different.
+
+First, in UNIX `/elektra/./../version` would be the same as `/version`, because `.` is just skipped and `..` removes `elektra`.
+But in Elektra, we evaluate `..` first and only then skip `.`.
+`/elektra/./../version` is the same as `/elektra/version`, because `..` remove `.`.
+
+The last example is not as different, in UNIX such paths would also refer to the same file.
+However, in some UNIX tools, a trailing slash alters the behaviour of the tool.
+In Elektra this is never the case.
+`/elektra/version/` and `/elektra/version` refer to the same Key and are always treated as the same Key Name.
+The only exception are the Root Keys.
+The Canonical Key Names for the Root Keys always end with a `/`.
+In fact, we will see [later](#4-valid-and-invalid-key-names), that removing the `/` makes the Key Name invalid.
+
+There also is a completely new addition in Elektra.
+Elektra has a notion of _Array Parts_.
+These are Key Name Parts that denote an array index.
+How exactly these work, will be explored [later](#4-valid-and-invalid-key-names).
+For now, we only need to know that they start with an `#` and their canonical form has `n` underscores followed by `n+1` digits.
+
+A few examples for Array Parts:
+
+| Non-Canonical    | Canonical           |
+| ---------------- | ------------------- |
+| `/elektra/#10`   | `/elektra/#_10`     |
+| `/elektra/#1234` | `/elektra/#___1234` |
+
+### 2.2. Other Escape Sequences
+
+We already know, that `/` and `\` have to be escaped in an Escaped Key Name.
+In addition two these to, there are a few more characters that have to be escaped.
+However, these additional characters may **only** be escaped at the start of a Key Name Part.
+
+The characters in question are: `.`, `#`, `%`, `@`.
+For example the Canonical Escaped Key Name for the Key with Namespace "Cascading" and Key Name Parts `elektra`, `.` and `version` is: `/elektra/\./version`.
+The `\.` is the escaped form of `.`.
+Using just `.` would make it a Non-Canonical Key Name, whose Canonical counterpart is `/elektra/version`, which does not have the wanted Key Name Part `.`.
+Similarly, `#10` would make a Key Name Non-Canonical, but `\#10` won't.
+Therefore, if you want a Canonical Key Name with a Key Name Part `#10` and not actually `#_10`,
+must use e.g. `/elektra/\#10/version`.
+
+<!-- TODO (kodebach): Unforunate escaping behaviour
+        Unfortunately this means that both `/elektra/.version` and `/elektra/\.version` have the same unescaped form (`elektra`, `.version`).
+        The same applies to `/elektra/%version` and `/elektra/\%version`.
+        All of these are, however, Canonical Key Names.
+        So the Key (CASCADING, `elektra`, `.version`) actually has two canonical forms....
+-->
```

Comment:
@markus2330 @mpranj As the comment says, I found this somewhat unfortunate behaviour. Not sure, what to do about it. We could say that `\.` is only a valid escape sequence, if the whole Key Name Part is `\.` or `\..`. I think that should probably be fine. For `\%` I'm not sure if that would work, because of the whole contextual value stuff that is so full of template code that I still don't understand it. I'm just very glad that it seems to still work apart from a tiny change related to the `default` namespace.

An alternative would be to canonicalize `\.version` and `\%version` into `.version` and `%version`, but again, not sure how the plays with the contextual value stuff. I think this will cause more problems, because AFAIK the contextual value stuff only uses the escaped name (and doesn't respect escaped slashes).

The best solution might be to ignore it for this PR and think of a fix before the 1.0 release.

Reply:
I am okay with removing the special meaning of the `.` in the beginning of key name parts (the whole "hiding" feature is obsolete).

The `%` is not related to contextual values, it represents empty key name parts. You can forget about contextual values, there is no need to escape them, as they do not appear anymore in libelektra-core (they are resolved in the layer above).

Reply:
> The % is not related to contextual values, it represents empty key name parts.

Is there actually any need for that? It really makes very little sense to have an empty key name part anywhere. Even if you want to use a user supplied value as a key name part, you could just add a prefix/suffix to avoid empty key name parts.

> You can forget about contextual values, there is no need to escape them, as they do not appear anymore in libelektra-core (they are resolved in the layer above).

They may be resolved at the layer above, but it is still relevant to all of Elektra that such keys can have a special meaning. That was the whole point of #3223. We **need** to define a set of reserved characters that plugins, libraries and applications can use for special purposes (ideally in a semi-regulated fashion like `METADATA.ini` for metadata). Otherwise an application might use the key `/somewhere/%user%` and because the user mounted a certain plugin this turns into e.g. `/somewhere/kodebach` without the application knowing.

That is also why I am so insistent on having an easy to check a key name part for special meaning (e.g. checking just the first or first few bytes). If a component of Elektra needs to check the whole key name to make sure the key name won't be interpreted in a special way by some other component of Elektra, the whole thing will get insanely complicated and extremely slow.

Reply:
> `%` Is there actually any need for that?

Yes, it is essential, otherwise keySetBaseName would not be able to work with any string. An empty string is as much string as any other. And yes, there are file formats that allow such things, e.g. TOML.

> you could just add a prefix/suffix

Of course you theoretically could always do millions of workarounds (practically it would not work: we would get millions of bugs in all storage plugins) but everybody still prefers if library functions like keySetBaseName just work without exceptions.

> They may be resolved at the layer above, but it is still relevant to all of Elektra that such keys can have a special meaning.

This is a wrong thought: only because something had a special meaning somewhere else, does not mean we need to do something about it. The % we get after the contextual evaluation do not have the contextual meaning anymore and it would be non-sense to give them a special treatment.

That `%` alone has a special meaning is just a coincident. (Like `.` and `.hidden` which is also simply a reuse of the same character.)

> because the user mounted a certain plugin this turns into

Contextual values unfortunately cannot be implemented as normal plugin as they happen at lookup.

> That is also why I am so insistent on having an easy to check a key name part for special meaning (e.g. checking just the first or first few bytes). If a component of Elektra needs to check the whole key name to make sure the key name won't be interpreted in a special way by some other component of Elektra, the whole thing will get insanely complicated and extremely slow.

We do not know if it would get slow. But yes, in general it fits better to Elektra if we have very short key part names which only have a single meaning (and a simple way to escape this meaning). The contextual values as implemented have been unnecessarily powerful but they are irrelevant in this escaping discussion anyway.

Reply:
> Yes, it is essential, otherwise keySetBaseName would not be able to work with any string. An empty string is as much string as any other.

That is **not** a reason IMHO. **We** define what `keySetBaseName` accepts and whether it makes sense that is should accept any string.

> And yes, there are file formats that allow such things, e.g. TOML.

This is slightly better argument but it is still beside the point.

Do you know of any instance where an application would use an empty key name part for its configuration where this cannot be easily avoided?

TOML, JSON, YAML and a million other formats might allow it. But if nobody actually uses it (because it is silly) it is still irrelevant to us.

> we would get millions of bugs in all storage plugins

AFAIK, there are just 2 plugins that can deal with any `KeySet` correctly: `quickdump` and (after #3507) `dump`. `mmapstorage` too, but only if absolute key names in the storage files are acceptable. Every other plugin runs into problems somwhere. Either the underlying storage format is limited or the plugin is buggy itself. In most cases, the limitation is that the storage file must be valid UTF-8 but Elektra's key names might not be.

> This is a wrong thought: only because something had a special meaning somewhere else, does not mean we need to do something about it.

The idea of Elektra is that an application is completely agnostic to the actual setup of the KDB. Therefore, if an application wants to use arbitrary key name parts (e.g. your example of WiFi SSIDs), then there has to be some way of guaranteeing that **no part** of Elektra no matter how obscure can interfere with this. Otherwise this misses the whole point of Elektra. The user should be in total control of their KDB and the application should just work.

> That `%` alone has a special meaning is just a coincident.

I know, and this entirely irrelevant.

> Contextual values unfortunately cannot be implemented as normal plugin as they happen at lookup.

Again, entirely irrelevant, as this was just an example.

Let's call it the `percent_user_percent` plugin. It's only job is replacing `%user%` with the current users name during `kdbGet` and `kdbSet`. This is a pretty stupid plugin, but there is no reason why this shouldn't be allowed.

Now someone else writes an application that does:
```c
Key * key = keyNew ("user/myapplication", KEY_END);
keySetBaseName (key, "%user%");
```

This is valid, because `keySetBaseName` accepts any string. The application then passes this key to `kdbSet`. The next time we start the application it calls `kdbGet` and does a `ksLookupByName (ks, "user/myapplication/%user%", 0)`.

Now we have a problem. If the user mounted `percent_user_percent`, the `ksLookupByName` won't work as expected.

I can give a million other examples like this. All of them just to show that we **need** some rules/restrictions for key names.

> We do not know if it would get slow.

Yes we do. It might be a minuscule difference for a single key name, but if for array parts alone there will be a noticeable difference at some point.

Currently this is the fastest way (that I know of) to check, if a single key name part `baseName` is a valid array part:

```c
int elektraArrayValidateBaseNameString (const char * baseName)
{
	const char * current = baseName;
	if (current == NULL || *current != '#')
	{
		return -1;
	}

	current++;

	if (*current == '\0')
	{
		return 0;
	}

	int underscores = 0;
	int digits = 0;

	while (*current == '_')
	{
		current++;
		underscores++;
	}

	while (isdigit ((unsigned char) *current))
	{
		current++;
		digits++;
	}

	bool underscoresCorrect = underscores == digits - 1;
	bool totalLengthCorrect = underscores + digits <= ELEKTRA_MAX_ARRAY_SIZE - 2;
	bool reachedEnd = *current == '\0' || *current == '/';

	if (underscoresCorrect && totalLengthCorrect && reachedEnd)
	{
		return underscores + 1;
	}

	return -1;
}
```

This assumes that `baseName` is an actual key name part that was validated by `elektraKeyNameValidate` at some point and not just an arbitrary string. For arbitrary strings you also need a `strncmp (&baseName[1 + underscores], "9223372036854775807", 19)`. It also only works for canonical key names.

Obviously `elektraKeyNameValidate` has to do all these checks, but it is really stupid that we need to do those checks again for already validated key name parts. If instead could just do `baseName[0] == '#'`, it would obviously be faster. How much faster might be unclear, but it is indisputable that a single byte comparison is faster than multiple loops and comparisons.

And this doesn't even take into account that because "`keySetBaseName` accepts any string" and the WiFi SSID example, you would technically have to check the escaped name as well to be entirely certain. Checking the escaped name is incredible complicated compared to even the above version of `elektraArrayValidateBaseNameString`, and definitely slower yet.

If you are unsure, why the escaped name has to be involved consider an application that uses keys of the form `/myapp/[WIFI-SSID]/autoconnecet` to store whether to automatically connect to a WiFi network. Since the WiFi-SSID is supplied by the user, it could be `#abc` or `#_12`. The application uses `keySetBaseName` to construct the keys, because it knows "`keySetBaseName` accepts any string". But how would `keySetBaseName` know that this time `#_12` should not be considered an array element when normally it should? It simply can't know, but let's assume you need to use some other function for array elements and `keySetBaseName` always escapes `#` (**not** the case right now). Even in that case, we will get the unescaped name parts `#abc` and `#_12`. To know that `#_12` is not supposed to be an array part, we would have to check the escaped name.

Alternatives:
1. (currently the case) `keySetBaseName` only escapes `#` when needed. In this case `#_12` will be treated as an array part and there is no way to detect that it shouldn't be.
2. There is no guarantee that `keySetBaseName` argument of `keySetBaseName` ends up being the last unescaped key name part. This breaks the expectation that `keySetBaseName (key, keyBaseName (key))` is a no-op on `key`, which could potentially cause **a lot** more trouble than restrictions on key names.
3. `keySetBaseName` doesn't accept any string. Not an option, because. Well, I don't know why, but you seem to be sure that it is not possible.

_**MOST OF ALL:**_ It might be still be fast enough to check the entire escaped and unescaped key name 10 times, but even if that were the case, it would still be slow, complicated and error-prone to write and maintain these checks. And because any plugin can currently claim special meaning for any key name part the whole exercise is irrelevant, because we don't even know what we have to check.

Reply:
Thank you very much, I really appreciate that you try to get the best Elektra possible. As the discussion is scattered around everywhere and it gets impossible for others to understand what the outcome is, I started to write decisions in #3514. Please comment, as far as possible, there directly on the text so that we can come to an conclusion.

> We define what keySetBaseName accepts and whether it makes sense that is should accept any string.

Of course, I started the base_name.md decision to discuss this.

> Do you know of any instance where an application would use an empty key name part for its configuration where this cannot be easily avoided?

No and I would also not consider it as best practice to do something like this. KeySets, however, are not only for applications but also to represent contents of configuration files. And to have bug-free fully-conforming storage plugins is an important goal 86ceebb9daa7a841da543d7e62a685d207f9d220

> TOML, JSON, YAML and a million other formats might allow it. But if nobody actually uses it (because it is silly) it is still irrelevant to us.

If this would be an argumentation about some concrete configuration parameters, I would fully agree. But we discuss here the possibility if any application, which might not even exist at the moment, could ever need a key part name which is empty. I do not know the answer. Fact is, that we support it at the moment and that configuration file formats require its support, too. So it must be an extremely good argument to drop such a support.

> The idea of Elektra is that an application is completely agnostic to the actual setup of the KDB.

Yes, this was an idea/goal but it did not work. See capabilities.md

> then there has to be some way of guaranteeing that no part of Elektra no matter how obscure can interfere with this

Yes, I agree: characters.md

> but there is no reason why this shouldn't be allowed.

It would be allowed in the decision with `@`.

> All of them just to show that we need some rules/restrictions for key names.

Of course we need to escape `@` if we do not want its meaning. You are totally right about this and this is a big oversight on my part.

> elektraKeyNameValidate

does not even need to be called in a mmap scenario, so it does not matter if this is slow. Furthermore, I highly doubt that iterating several times over the key names will be noticeable compared to everything else going on as validation only happens at key construction. So these performance discussions are, as long as there is no benchmark showing a slowdown by a factor two of the whole `kdbGet` in a realistic scenario, simply irrelevant.

> #_12 should not be considered an array element when normally it should?

No, it is only an array element if the direct parent key has meta-data array, see array.md. So you implemented this already correctly afaik.

> This breaks the expectation that keySetBaseName (key, keyBaseName (key))

Yes, of course we want this. f8822fc4b47023a6732ae2f47175ee732a734755

> Well, I don't know why, but you seem to be sure that it is not possible.

:wink: It is my experience: we had many problems in storage plugins and after `keySetBaseName` started accepting everything, these problems disappeared.

> complicated and error-prone to write and maintain these checks.

Yes but much easier to maintain these checks at one place (and not duplicated for every storage plugin).

Reply:
> No, it is only an array element if the direct parent key has meta-data array, see array.md. So you implemented this already correctly afaik.

1. This is not possible for arrays in metadata. Metakeys cannot have metadata.
2. Relying on another key to decide the interpretation of a key is ridiculous, IMHO. That would be like saying this file is a JSON file unless the directory is a owned by root than it is a YAML file.

> It is my experience:  we had many problems in storage plugins and after `keySetBaseName` started accepting everything, these problems disappeared.

There will always be problems in storage plugins, especially since most of them are unmaintained. Just because you had bad experiences, doesn't mean it is impossible to provide an easy to use API for storage plugins in which `keySetBaseName` does not accept anything.

> Yes but much easier to maintain these checks at one place (and not duplicated for every storage plugin).

`storage` library...

Reply:
>  This is not possible for arrays in metadata. Metakeys cannot have metadata.

I do not think that this is a problem as meta-arrays are pure convention and not a real thing (e.g. that are somehow serialized differently). I added this assumption in e533275e7f1bd90b5954051f536ba3b1301cb521

> Relying on another key to decide the interpretation of a key is ridiculous, IMHO. That would be like saying this file is a JSON file unless the directory is a owned by root than it is a YAML file.

Why do you think so? I think the other way: I would like to avoid having any semantics in the key names. For me semantics belong to metadata. Metadata is specifically designed to give keys semantics. For me the `#...` syntax is pure convention, it does not give any meaning and can be used anywhere where you want to count up something (also for non-arrays).

> There will always be problems in storage plugins, especially since most of them are unmaintained.

Which storage plugin is unmaintained? Imho only `ini`, which we will remove in #3491. We need to have realistic expectations of what a maintainer can do. In other free-software projects, like Linux, it is also common practice that big changes include the modification of all drivers, not that individual maintainers do it.

> Just because you had bad experiences, doesn't mean it is impossible to provide an easy to use API for storage plugins in which keySetBaseName does not accept anything.

I also did not claim that.

Reply:
> I do not think that this is a problem as meta-arrays are pure convention and not a real thing (e.g. that are somehow serialized differently).

That is just because, there is no storage plugin yet that supports metadata and also has a native array format.

Let's imagine `son`. It's like `ni` but for JSON. For this `ni` file:

```ini
my/key = 5

[my/key]
my/meta/#0 = a
my/meta/#1 = b
```

The `son` plugin could produce this JSON:

```json
{
  "keys": {
    "my/key": 5
  },
  "meta": {
    "my/key": {
      "my/meta": [ "a", "b" ]
    }
}
```

> I would like to avoid having any semantics in the key names. For me semantics belong to metadata. Metadata is specifically designed to give keys semantics. For me the `#...` syntax is pure convention, it does not give any meaning and can be used anywhere where you want to count up something (also for non-arrays).

This interpretation does make sense, yes. But then the metadata would have to be on the array elements. IMO the semantics of a `Key` should not change based on context.

Just think about a cascading key `/myarray/#0`. Where does the `array` metakey have to exist to make this an array element? Do you have to lookup `/myarray/#0` to get e.g. `user:/myarray/#0` and then `user:/myarray` has to have the metadata? Do we just do a lookup with `/myarray` and if `user:/myarray` doesn't exist, but `system:/myarray` has the `array` metadata its enough, and `user:/myarray/#0` is still an array element? What about the `spec` namespace?

The idea with the parent key is nice, but it is much more complicated and requires a lot more definitions than just "`#...` is an array element".

Reply:
> there is no storage plugin yet that supports metadata and also has a native array format.

Yes, this is not covered by this decision. Obviously, you can simply guess that it is an array by the structure. This was what we were doing for normal arrays quite some time. This works quite well, except for empty arrays.

> But then the metadata would have to be on the array elements. IMO the semantics of a Key should not change based on context.

I agree it would be more beautiful if every key is standalone. Any suggestion? Adding `array/element` with a pointer to the parent to every element sounds very expensive (memory-wise), even if the meta-data is copied.

> Just think about a cascading key /myarray/#0. Where does the array metakey have to exist to make this an array element?

It does not work the reverse way. Where would this be needed?

> Do we just do a lookup with /myarray

Exactly, I added the description of that 1c729469fddb5125c0621ff369c87447b59aafb7

> and if user:/myarray doesn't exist, but system:/myarray has the array metadata its enough, and user:/myarray/#0 is still an array element?

I gave an example.

> What about the spec namespace?

What exactly do you mean? This is unaffected by the decision.

> The idea with the parent key is nice, but it is much more complicated and requires a lot more definitions than just "#... is an array element".

For most applications this will be true anyway. I do not think that many applications will abuse `#...` for anything else. The metadata `array` was mostly needed for empty arrays. I don't see any situation where it would be a problem to add the metadata (except of meta-arrays of course but this is a different topic).

Reply:
> Adding array/element with a pointer to the parent to every element sounds very expensive (memory-wise), even if the meta-data is copied.

I don't think `array/element` would need a value. Just `array/element = ` should be enough. Its obvious that the parent key is the array container. That way we could even reuse a single `Key` instance across the whole KDB.

> Re: array lookup

I commented on #3514.

> > The idea with the parent key is nice, but it is much more complicated and requires a lot more definitions than just "#... is an array element".
> 
> For most applications this will be true anyway. I do not think that many applications will abuse `#...` for anything else. The metadata `array` was mostly needed for empty arrays. I don't see any situation where it would be a problem to add the metadata (except of meta-arrays of course but this is a different topic).

You misunderstood. I didn't mean the application has to define more things. I meant it requires more definitions from us to specify how Elektra works.

Another somewhat confusing thing to me is: With #3447 the core definitely knows about array elements, since it adds the underscores when needed. But at the same time we kind of pretend the core doesn't know about array elements, because "arrays are just a convention".

Reply:
> I don't think `array/element` would need a value. Just `array/element = ` should be enough. Its obvious that the parent key is the array container. That way we could even reuse a single `Key` instance across the whole KDB.

Actually that doesn't work. We could have a structure like this:

```
myarray/#0/x = 1
myarray/#0/y = 2
```

Here we would have to add `myarray/#0` just to contain the `array/element =` metadata. That is not worth the additional memory. We would also still need a lookup to find out if `myarray/#0/x` is part of an array.

However, such a structure is also problematic with the current `array` metadata. You need to do a lot of work to find out, whether `#0` is actually an array element. I now remember that this is actually the reason why I originally wanted to make it so that `#` always indicates array elements.

Reply:
Yes, I agree there is a lot of specification missing, also to clarify what an "array element" is. This huge PR, however, is not the right place for such a discussion. :wink: